From: Samuel Holland <samuel@sholland.org>
Date: Wed, 12 Jan 2022 00:39:37 -0600
Subject: spi: gpio: Yes, we do need delays when bitbanging

Because 10 kHz is the fastest speed the ws8100 (firmware) can handle.

Also, a large delay is needed after asserting CS. "CS" handling should
probably be moved entirely to the ws8100 driver.

Signed-off-by: Samuel Holland <samuel@sholland.org>
Signed-off-by: Maximilian Weigand <mweigand@mweigand.net>
---
 arch/arm64/boot/dts/rockchip/rk3566-pinenote.dtsi | 1 +
 drivers/spi/spi-gpio.c                            | 3 ++-
 2 files changed, 3 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/boot/dts/rockchip/rk3566-pinenote.dtsi b/arch/arm64/boot/dts/rockchip/rk3566-pinenote.dtsi
index 2c877fe..2f7b47a 100644
--- a/arch/arm64/boot/dts/rockchip/rk3566-pinenote.dtsi
+++ b/arch/arm64/boot/dts/rockchip/rk3566-pinenote.dtsi
@@ -153,6 +153,7 @@ bluetooth@0 {
 			reset-gpios = <&gpio3 RK_PC0 GPIO_ACTIVE_HIGH>;
 			spi-3wire;
 			spi-max-frequency = <10000>;
+			spi-cs-inactive-delay-ns = <7000000>;
 		};
 	};
 
diff --git a/drivers/spi/spi-gpio.c b/drivers/spi/spi-gpio.c
index 4f192e0..1b0380b 100644
--- a/drivers/spi/spi-gpio.c
+++ b/drivers/spi/spi-gpio.c
@@ -6,6 +6,7 @@
  * Copyright (C) 2017 Linus Walleij
  */
 #include <linux/gpio/consumer.h>
+#include <linux/delay.h>
 #include <linux/kernel.h>
 #include <linux/mod_devicetable.h>
 #include <linux/module.h>
@@ -113,7 +114,7 @@ static inline int getmiso(const struct spi_device *spi)
  * reaching even one Mbit/sec (except when we can inline bitops), so for now
  * we'll just assume we never need additional per-bit slowdowns.
  */
-#define spidelay(nsecs)	do {} while (0)
+#define spidelay(nsecs)	ndelay(nsecs)
 
 #include "spi-bitbang-txrx.h"
 
